### PR TITLE
[shardformer] support DDP in HybridPlugin/add tp+dp tests

### LINKED ISSUE #4422: [shardformer] support tp+dp

### PR DESCRIPTION
**Pull request by Fridge003**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

closed #4422

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

- Support tp+dp for shardformer models, where dp is optimized using ddp
- Add relevant tests

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/booster/plugin/hybrid_parallel_plugin.py [hunk 1]
```diff
@@ -45,7 +47,23 @@ def __init__(self, module: Module, precision: str, shard_config: ShardConfig, dp
...
         else:
             module = module.cuda()    # train without AMP
-        # TODO(ver217): support TP+DP
+
+        if use_ddp:
+
+            ddp_config = dict(process_group=dp_group,
+                              broadcast_buffers=True,
+                              bucket_cap_mb=25,
+                              find_unused_parameters=False,
```
Comment (ver217) [inline colossalai/booster/plugin/hybrid_parallel_plugin.py:56]: `find_unused_parameters=False` and `static_graph=False` may lead to correctness issue when there are unused params. This is commonly observed when layer drop is used.

File: tests/test_shardformer/test_model/_utils.py [hunk 2]
```diff
@@ -246,3 +248,17 @@ def check_grad(org_model: Module,
         assert torch.allclose(
             org_grad.float(), shard_grad.float(), rtol=rtol, atol=atol
         ), f"error attribute '{suffix}', orgin model grad is not equal to shard model grad\n{org_grad}\n{shard_grad}"
+
+
+def unwrap_model(module: Module,
+                 base_model_class_name: Optional[str] = None,
+                 base_model_attribute_name: Optional[str] = None):
+    if isinstance(module, HybridParallelModule):
```
Comment (ver217) [inline tests/test_shardformer/test_model/_utils.py:256]: Model wrapper class already has `unwrap()` method. I think you should overwrite this method in `HybridParallelModule`.

Comment (Fridge003) [reply tests/test_shardformer/test_model/_utils.py:256]: This util function is only for the convenience of testing, so I'm putting it under testing folder.

Comment (Fridge003) [reply tests/test_shardformer/test_model/_utils.py:256]: This util function is only for testing usage, but it makes sense to overwrite the unwrap() method in HybridParallelModule to deal with DDP cases.

File: colossalai/booster/plugin/hybrid_parallel_plugin.py (hunk 1 above)
Comment (Fridge003) [reply colossalai/booster/plugin/hybrid_parallel_plugin.py]: modified

**Review by ver217 — APPROVED** [review]

**Pull request merged**